[DRAFT] Support for custom uniform buffers and shader storage buffers in GDShaders

Fluxx7 opened the pull request:
Right now, the only ways to pass data to a GDShader is through uniforms or textures. Uniforms are solidly limited by the capacity of the single UBO dedicated to material uniforms and they don't support structs, while textures can be confusing to work with and require wasting performance encoding and decoding them. This PR adds syntax to GDShaders for both uniform and storage buffers, as well as methods to set them from a script. 

An example of the syntax I have implemented is:
`buffer uniform restrict my_buf {`
`    float bufFloat; `
`}(bindname = myBuffer, format = std140)`
The buffer value could then be read in the shader as `my_buf.bufFloat`. 

To set the values in the buffer from a script, you would use 
`myshadermat.set_shader_buffer("myBuffer", 
    { "bufFloat": 0.1 })`
You can also directly set the buffer values with a PackedByteArray if you prefer via `set_shader_buffer_raw()`
 
Right now, the buffers work properly and can be assigned to and read from without issue. Structs can be passed through buffers, as well as arrays (and arrays of structs). Once everything is feature complete, I'll clean up the commits and reformat this PR to be better structured, but I figured it was worth sharing as is.

Closes: https://github.com/godotengine/godot-proposals/issues/7516
Also closes https://github.com/godotengine/godot-proposals/issues/6989

If you want to quickly see a project implementing it, I have published my water rendering project that I've been using for testing so far, which can be accessed [here](https://github.com/Fluxx7/SSBO-testing). I'm going to continue to update this project to try and make it a proper demo for the various options, it is very WIP right now so the code is sort of a nightmare.

Completed work (somewhat summarized):
- [X] Implement syntax for declaring both UBOs and SSBOs
- [X] Add support to compiler for buffer instantiation + buffer value accesses
- [X] Properly allocate and bind buffers to material uniform set
- [X] Implement API for scripts to set buffer data
- [X] Implement completion options for buffers, buffer fields, and buffer formats
- [X] Support using both named and unnamed buffers
- [X] Support both std140 and std430 for storage buffers
- [X] Abstract the process of setting buffer values from scripts

Remaining work in rough order of priority:

- [ ] Ensure support for buffers being shared between shaders
- [ ] Fix unsized storage buffer arrays (I know what's going wrong but it's gonna be a pain to fix)
- [ ] Add support for buffers as a global parameter
- [ ] Fix completion options for buffer IO qualifiers
- [ ] Add parser checks to ensure that writeonly buffers cannot be read from (not sure writeonly buffers are even necessary tbh)
- [ ] Validate that storage buffer values can be changed by the shader (again, not sure this is really useful but no reason not to have it)
- [ ] Write unit tests
- [ ] Write documentation
- [ ] Final optimizations

Fluxx7 committed:
laying groundwork + figuring out codebase

Fluxx7 committed:
began work on compiler end

Fluxx7 committed:
syntax parses, rendering thread does not like

Fluxx7 committed:
SL supports buffers now

Haven't tested any edge cases, but uniform buffers specifically can be created.

I now need to actually instantiate those buffers and let the user change the values in scripts

Fluxx7 committed:
Uniform buffers work now

Fluxx7 committed:
storage buffers work, ironing out issues

Fluxx7 committed:
formatting fixes

Fluxx7 committed:
Fixed compilation errors with unnamed buffers

Fluxx7 committed:
implemented better buffer value setting API

currently only supports floats, vector2s, and structs/arrays comprised only of those two

Fluxx7 committed:
set_shader_buffer supports more types now

Fluxx7 committed:
all API functions implemented + more type support

Redwarx008 commented:
A method is needed to be able to receive from an existing buffer created by a RenderingDevice.
Just like Texture2Drd:
``` c#

var buffer = new BufferRD();
buffer.Rid = rd.StorageBufferCreate();
ShaderMaterial.SetShaderBuffer("name", buffer );
```

Fluxx7 commented:
I was trying to think of how to approach that, BufferRD looks like a good solution

Fluxx7 commented:
To be honest, I don't fully understand what you're asking for in relation to your code, but the goal is to support sharing buffers between GDShaders and compute shaders, provided the buffer is created on the main RenderingDevice. My test project right now has to read back the values from the compute shader and then set the buffer that is read in both the vertex and fragment shader to those values, but once this goal is completed, the buffer would be bound to both in the same way a TextureRD can be bound to both, so no read back would be needed

Fluxx7 commented:
I could pretty easily expose API to directly get or set the buffer RID for the GDShaders, but I'm trying to make the API as user-friendly as possible, and juggling RIDs in scripts always sucks imo. I could add those functions as an option I guess, but setting the buffer RID could make some of the other API stuff not work properly or more difficult to implement properly, so I'm hesitant to add that. Hopefully direct RID access won't be needed once I finish this refactor and get the API to work properly

Deltt commented:
Just stumbled on this, can't say how happy I am this seems to be actively worked on.

On the matter of juggling RIDs, I personally find it pretty user friendly to manage them and would prefer that over a more simple API that potentially would be less flexible in exchange for being more user friendly.

Fluxx7 committed:
working on BufferRD refactor

BufferRD is similar to the TextureRD objects and should allow for safe API that's still flexible. I'm also trying to move some of the low-level stuff to RenderingDevice so that buffers can be shared to compute shaders properly

Fluxx7 committed:
Merge remote-tracking branch 'upstream/master' into gdshader_storage_buffers

jeffercizer commented:
any luck on the refactor, I also need SSBOs to be shareable between compute and vertex shaders for my project

Fluxx7 commented:
I haven't made any progress for a while, I had a very busy year of university lol. Now that I have more free time, I'm going to try and get this PR fully working asap, hopefully in time for 4.8 if no major changes need to be made after the PR is ready
